operators/test/e2e/cmd/run/run.go

```diff
@@ -0,0 +1,561 @@
+		if err := enc.Encode(h.testContext); err != nil {
+			return errors.Wrap(err, "failed to encode test context")
+		}
+	}
+
+	return nil
+}
+
+func (h *helper) createE2ENamespaceAndRoleBindings() error {
+	log.Info("Creating E2E namespace and role bindings")
+	return h.kubectlApplyTemplate("config/e2e/rbac.yaml", h.testContext, true)
+}
+
+func (h *helper) installCRDs() error {
+	log.Info("Installing CRDs")
+	crds, err := filepath.Glob("config/crds/*.yaml")
+	if err != nil {
+		return errors.Wrap(err, "failed to list CRDs")
+	}
+
+	for _, crd := range crds {
+		log.V(2).Info("Installing CRD", "crd", crd)
+		if err := h.kubectlApplyTemplate(crd, h.testContext, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *helper) deployGlobalOperator() error {
+	log.Info("Deploying global operator")
+	return h.kubectlApplyTemplate("config/e2e/global_operator.yaml", h.testContext, true)
+}
+
+func (h *helper) deployNamespaceOperators() error {
+	log.Info("Deploying namespace operators")
+	return h.kubectlApplyTemplate("config/e2e/namespace_operator.yaml", h.testContext, true)
+}
+
+func (h *helper) deployTestJob() error {
+	log.Info("Deploying e2e test job")
+	return h.kubectlApplyTemplate("config/e2e/batch_job.yaml", h.testContext, true)
+}
+
+func (h *helper) runTestJob() error {
+	if h.setupOnly {
+		log.Info("Skipping tests because this is a setup-only run")
+		return nil
+	}
+
+	client, err := h.createKubeClient()
+	if err != nil {
+		return errors.Wrap(err, "failed to create kubernetes client")
+	}
+
+	// start the event logger to log all relevant events in the cluster
+	stopChan := make(chan struct{})
+	eventLogger := newEventLogger(client, h.testContext, h.eventLog)
+	go eventLogger.Start(stopChan)
+
+	// launch the test job and wait for it to finish
+	err = h.monitorTestJob(client)
+	close(stopChan)
+
+	if err != nil {
+		h.dumpEventLog()
+		return errors.Wrap(err, "test run failed")
+	}
+
+	return nil
+}
+
+func (h *helper) createKubeClient() (*kubernetes.Clientset, error) {
+	// load kubernetes client config
+	var overrides clientcmd.ConfigOverrides
+	var clientConfig clientcmd.ClientConfig
+
+	if h.kubeConfig != "" {
+		kubeConf, err := clientcmd.LoadFromFile(h.kubeConfig)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to load kubeconfig")
+		}
+
+		clientConfig = clientcmd.NewDefaultClientConfig(*kubeConf, &overrides)
+	} else {
+		loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+		clientConfig = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &overrides)
+	}
+
+	// create the kubernetes API client
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create kubernetes client config")
+	}
+
+	return kubernetes.NewForConfig(config)
+}
+
+// monitorTestJob keeps track of the test pod to determine whether the tests failed or not.
+func (h *helper) monitorTestJob(client *kubernetes.Clientset) error {
```

Comment:
I don't have a strong opinion on jobs vs pods but I think that the amount of work wouldn't be much less because we have to wait for the pod to start in either case.